### ISSUE TITLE
[Android] Adding the Spinner Option to Business Hours Editor

### ISSUE THREAD

**Issue by melonSkin76** [organicmaps/organicmaps#2534]
Hi everyone,

Recently, I found issue #2482 to be interesting. So, I spent some time playing with Android Studio to solve it. Here is a simple demo of my solution:
![timepicker_demo](https://user-images.githubusercontent.com/53326572/167349973-649800b1-208d-4111-b11a-198e73716334.gif)

I added a new button "Timepicker Mode" to the Business Hours Editor Page (android/res/layout/item_timetable.xml). And I also created a new pop-up dialog (SelectTimepickerModeFragment) to let the users select their preferred timepicker mode. Please check the details in my commit to branch [rq-select-timepicker] of my [forked repository](https://github.com/melonSkin76/rq_organicmapsTest/commit/a28c86d4423c31670aa6c03cfa58f2dd13b8566a).

One of the challenges I encountered was that the [TimePicker  Class](https://developer.android.com/reference/android/widget/TimePicker) does not have a set method for its attribute "android:timePickerMode" and after a couple of searches I found it difficult to dynamically manipulate the attribute of the xml element. So, my solution this is to add a alternative version of "android/res/layout/fragment_timetable_picker.xml", which is called "fragment_timetable_picker_spinner.xml." Their only differences are the ids and the values of timePickerMode. In addition, I added some if statements to "android/src/com/mapswithme/maps/editor/HoursMinutesPickerFragment.java," which directly inflates the timepicker dialog. As a result, HoursMinutesPickerFragment will inflate different XMLs based on the users' choices.

Another problem I had was how the data communications between fragments should be done, i.e. how I can pass the timepicker's mode info from my new dialog to the actual timepicker dialog. My solution to this is to mimic the existing interaction logic between "HoursMinutesPickerFragment" and "SimpleTimetableAdapter." I made an listener interface for the new pop-up dialog (SelectTimepickerModeFragment) and then let SimpleTimetableAdapter to implement the listener. Then after the users choose the timepicker's mode, SimpleTimetableAdapter will pass its listened information to HoursMinutesPickerFragment through an existing method HoursMinutesPickerFragment.pick().

The two paragraphs above provide a brief explanation of how I solve the issue. I have not created a pull request yet because I really need your opinions on my changes.  I am kind of new to this project so I just want to make sure these changes will not create unexpected errors. I have tested this manually on my Android Emulators (API31 x86_64). And by far I have not seen any apparent issues. But actually there's a couple of holes I can think of: 1. The button for selecting modes is just a casual copy of the elements in the same layout, which needs to be designed more carefully to fit the app's style requirements. 2. The style of the pop-up dialog also needs some extra work, e.g. the dark theme is not supported.

Please let me know if you have any problems or suggestions!

**Comment by endim8** [comment]
very nice! it'd be nice if the button to switch mode was embedded in the popup, something like this: 
<img width="345" alt="Screenshot 2022-05-09 at 17 29 28" src="https://user-images.githubusercontent.com/26939824/167456092-8537edff-b379-4a5a-9be9-2071328deb95.png">

**Comment by biodranik** [comment]
Hi and thanks for your work and for the details!
I would suggest:
1. Use a visually clickable button/label that changes its text (clock/spinner) on click, without any popups (and save its value in settings, of course)
2. Try to simplify the code as much as possible by avoiding copy-paste in XML (if it's possible, of course). And also when passing values back. Please, fewer patterns and fewer listeners. Less code is better!

**Comment by dbf256** [comment]
There is another mode switch already, maybe combine them together?

![image](https://user-images.githubusercontent.com/2550295/167714294-b726c170-6f96-4246-88c7-4f707d28cc2a.png)

**Comment by melonSkin76** [comment]
Thank you all for the advice! @dbf256 I didn't even notice there was a button there... But still, this small keyboard button does not lead us towards a "spinner." Instead, the timepicker is switched to a mode that user can directly input the values like this
![image](https://user-images.githubusercontent.com/53326572/167725016-17991038-2710-4114-87e5-b3f770f7d34e.png)
But the problem is how I can manipulate this small button's functionality. I checked this in the layout inspector but this button seems to be a built-in feature of the TimePicker element: 
![image](https://user-images.githubusercontent.com/53326572/167726625-7dbe1cb3-e654-4023-8e26-c12db581e649.png)
In other words, I cannot find the id and onclick of this small button. So, maybe it is simpler to just add another visually clickable button like @biodranik and @endim8 mentioned before?

**Comment by biodranik** [comment]
Yes, add a button with a clear and changeable text depending on the next mode.

**Comment by melonSkin76** [comment]
I'm sorry for the delay. But I have finished a version with no popups and a clear text button (the corresponding listeners have also been removed). See the details in the branch of my [forked repository](https://github.com/melonSkin76/rq_organicmapsTest/commit/bcb5d0b857058c1f6e9d738a22ab91c320ba985a). Below is a simple demo:
![timepicker2_demo](https://user-images.githubusercontent.com/53326572/168695749-cb58b7a9-22f3-4dd4-9161-518d3129aff0.gif)
There is still one tiny problem that I have not managed to resolve -- after I click OK when the time period has has been picked, the button and the picker seem to be "reset" to the previous state. I set up several breakpoints (mainly around the text change operations) and did some debugging to find out the reasons behind this. And it turned out that no Button.setText() was called after I click OK in the picker. So, maybe there are view holder overlays where the bottom one is shuffled to the top after each time pick? (I studied the adapter and related fragment code for a while but I still not quite sure about the mechanisms) Nevertheless, this is just a small issue and it doesn't really affect the user experiences from my perspective.

I also made some attempts to add an extra button to the timetable picker dialog (fragment_timetable_picker.xml and HoursMinutesPickerFragment.java) just like @endim8 mentioned. As far as I am concerned, there are two ways to do this:
1. Keep the timepicker alert dialog (HoursMinutesPickerFragment) and simply replace its current view by a newly created view. I had some investigation on this and it turned out that the method "setView()" will not work after the dialog has been showed (I learned this from the reminder provided by Android Studio IDE). Another method "setContentView()," according to my tests, will replace the entire dialog and remove the buttons (positive, negative, and neutral). Also, according to the post [here](https://stackoverflow.com/questions/14751720/is-it-possible-to-replace-an-alertdialogs-view-after-the-dialog-is-shown), a more applicable way might be dismiss the dialog and build a new one.
2. Dismiss the current timepicker dialog and create a new one that is inflated differently. I think I am able to do this but this way may require extra message passing between HoursMinutesPickerFragment and SimpleTimetableAdapter. In other words, adding an extra interface of listener to the timepicker dialog seems inevitable because the SimpleTimetableAdapter need to know whether the change mode button has been clicked once the dialog is dimissed. But this is against the fewer listener principle presented by @biodranik.

So, is it okay to just have a button on the editor page like my demo above? I am looking forward to your suggestions.
